x/bank/errors.go

```diff
@@ -1,57 +1,41 @@
-//nolint
 package bank
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	cmn "github.com/tendermint/tmlibs/common"
 )
 
-// Coin errors reserve 100 ~ 199.
-const (
-	CodeInvalidInput  sdk.CodeType = 101
-	CodeInvalidOutput sdk.CodeType = 102
-)
-
-// NOTE: Don't stringer this, we'll put better messages in later.
-func codeToDefaultMsg(code sdk.CodeType) string {
-	switch code {
-	case CodeInvalidInput:
-		return "Invalid input coins"
-	case CodeInvalidOutput:
-		return "Invalid output coins"
-	default:
-		return sdk.CodeToDefaultMsg(code)
-	}
-}
-
 //----------------------------------------
 // Error constructors
 
+// error when provided inputs are invalid
 func ErrInvalidInput(msg string) sdk.Error {
-	return newError(CodeInvalidInput, msg)
+	return sdk.ErrGeneric(msg, []cmn.KVPair{
+		{[]byte("module"), []byte("bank")},
```

Comment:
not thrilled about this API. some thoughts:

- can we have a convention so all Err funcs within a module inherit the `module=xxx` without having to include the key every time ? Even if it's just defining a `var moduleTag KVPair` and using that everywhere instead of typing it out each time
- these will almost universally be strings. can we shield the bytes from the user somehow? a `map[string]string` might be nice but then we need to be sure to sort it . Note this relates to a recent fiasco in Tendermint, please see https://github.com/tendermint/tendermint/issues/1369#issuecomment-376852991 - would be great to get your input there too

Reply:
We shouldn't be using the tag system for this.  For one, all tags get persisted forever.  We should ideally use *no* tags, since using tags are expensive.

I've refactored the SDK error into tmlibs, with more modifications.  See here:
https://github.com/tendermint/tmlibs/blob/develop/common/errors.go

To figure out where the error came from, we should be using the Trace() function.  (Not the Stactrace() function by default, as that is expensive.  Trace() is a more lightweight alternative designed for figuring out where an error came from.  Generally you want this whenever a function can return an error from more than 1 source.  Then, in that function, you want to use Trace() to tell the programmer which path the error came from.  It's something that could be automated in the future, e.g. in Go2.

So I would do this:  Step 1:  convert SDKError to use tmlibs/common/Error.  It would embed NewError as perhaps an anonymous exposed field, but SDKError of course would need to support the Code as well.  SDKError can support tags too but we shouldn't be encouraging the use of tags for performance reasons, or more specifically because the code system was designed so that ideally we don't use maps or tags for errors.

Step 2: In the SDK, store the stringified form of SDKError (which includes all trace and stacktrace info from cmn.Error, and *append* it to SDKError.Log.  The Log is where all implementation-specific messages go.  Imagine if someone were to implement the Cosmos SDK in Rust.  The trace info will be different... Everything else should be deterministic except the SDKError.Log, so trace info must go there.  It's OK to do naive concatenation, like append a newline and maybe a visual separator, and more junk.

Reply:
ABCI also includes the "info" nondeterministic ReponseDeliverTx field.  
https://github.com/tendermint/abci/blob/develop/types/types.proto#L163

* Log should include the detailed output of the error including trace and stacktrace and all that, and also whatever the user wanted to log if any (and if we want to support that).
* Info should include a non-deterministic message of the error minus trace and stacktrace.  The message must correspond to the Code.
* Code should be convertible to a sensible string to display.  Clients can display Info as the primary error to display, or, it can use the Code and derive the Info message.  Maybe it makes sense to use the tags to format the message as well, but each Code would have to have a defined set of parameters.  For example,

```
CodeInvalidInput  sdk.CodeType = 101	
Default message for this code is "Invalid input coins".
We could change this to "Invalid input coins at index %v", and have in the tag, "index"="3".
These parameters would need to be defined in the comments for that Code, and all errors that use that Code should include these parameters.
func ErrNoInputs(index int) sdk.Error {	... }
```